encoding/ssz/query/proof_collector.go

```diff
@@ -0,0 +1,700 @@
+package query
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/bits"
+	"reflect"
+	"runtime"
+	"slices"
+	"sync"
+
+	"github.com/OffchainLabs/go-bitfield"
+	"github.com/OffchainLabs/prysm/v7/container/trie"
+	"github.com/OffchainLabs/prysm/v7/crypto/hash/htr"
+	ssz "github.com/OffchainLabs/prysm/v7/encoding/ssz"
+	fastssz "github.com/prysmaticlabs/fastssz"
+	"github.com/prysmaticlabs/gohashtree"
+)
+
+// ProofCollector collects sibling hashes and leaves needed for Merkle proofs.
+//
+// Multiproof-ready design:
+// - requiredSiblings/requiredLeaves store which gindices we want to collect (registered before merkleization).
+// - siblings/leaves store the actual collected hashes.
+//
+// Concurrency:
+// - required* maps are read-only during merkleization.
+// - siblings/leaves writes are protected by mu.
+type ProofCollector struct {
+	mu sync.Mutex
+
+	// Required gindices (registered before merkleization)
+	requiredSiblings map[uint64]struct{}
+	requiredLeaves   map[uint64]struct{}
+
+	// Collected hashes
+	siblings map[uint64][32]byte
+	leaves   map[uint64][32]byte
+}
+
+func NewProofCollector() *ProofCollector {
+	return &ProofCollector{
+		requiredSiblings: make(map[uint64]struct{}),
+		requiredLeaves:   make(map[uint64]struct{}),
+		siblings:         make(map[uint64][32]byte),
+		leaves:           make(map[uint64][32]byte),
+	}
+}
+
+func (pc *ProofCollector) Reset() {
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	pc.requiredSiblings = make(map[uint64]struct{})
+	pc.requiredLeaves = make(map[uint64]struct{})
+	pc.siblings = make(map[uint64][32]byte)
+	pc.leaves = make(map[uint64][32]byte)
+}
+
+// AddTarget register the target leaf and its required sibling nodes for proof construction.
+// Registration should happen before merkleization begins.
+func (pc *ProofCollector) AddTarget(gindex uint64) {
+	// Lock safe just in case the collector is re-used.
+	pc.mu.Lock()
+	defer pc.mu.Unlock()
+
+	pc.requiredLeaves[gindex] = struct{}{}
+
+	// Walk from the target leaf up to (but not including) the root (gindex=1).
+	// At each step, register the sibling node required to prove inclusion.
+	nodeGindex := gindex
+	for nodeGindex > 1 {
+		siblingGindex := nodeGindex ^ 1 // flip the last bit: left<->right sibling
+		pc.requiredSiblings[siblingGindex] = struct{}{}
+
+		// Move to parent
+		nodeGindex /= 2
+	}
+}
+
+// toProof converts the collected siblings and leaves into a fastssz.Proof structure.
+// Current behavior expects a single target leaf (single proof).
+func (pc *ProofCollector) toProof() *fastssz.Proof {
```

Comment:
```suggestion
func (pc *ProofCollector) toProof() (*fastssz.Proof, error) {
```

Current `toProof` method might hide the error like "Missing sibling", so I'd say we can let the caller decide when unwanted `*fastssz.Proof`is returned.

Reply:
Correct! I better return an error in this `toProof` before if the Sibling was not collected rather than hiding it up until the verifier attempts to verify the proofs